google/cloud/spanner/row.h

```diff
@@ -0,0 +1,179 @@
+ *
+ * Example:
+ *
+ *     spanner::Row<bool, std::int64_t, std::string> row;
+ *     row = spanner::MakeRow(true, 42, "hello");
+ *     static_assert(row.size() == 3, "Created with three types");
+ *
+ *     // Gets all values as a std::tuple
+ *     std::tuple<bool, std::int64_t, std::string> tup = row.get();
+ *
+ *     // Gets one value at the specified column index
+ *     std::int64_t i = row.get<1>();
+ *
+ *     // Gets the values from the specified columns (any order), and returns a
+ *     // tuple. Since C++17, this tuple will work with structured bindings to
+ *     // allow assigning to individually  named variables.
+ *     auto [name, age] = row.get<2, 1>();
+ *
+ */
+template <typename... Types>
+class Row {
+  template <std::size_t I>
+  using ColumnType = typename std::tuple_element<I, std::tuple<Types...>>::type;
+
+ public:
+  /// Regular value type, supporting copy, assign, move, etc.
+  Row() = default;
+  Row(Row const&) = default;
+  Row& operator=(Row const&) = default;
+  Row(Row&) = default;
+  Row& operator=(Row&&) = default;
+
+  /**
+   * Constructs a Row from the specified values.
+   *
+   * See also the `MakeRow()` helper function template for a convenient way to
+   * make rows without having to also specify the value's types.
+   */
+  template <typename... Ts>
+  explicit Row(Ts... ts) : values_(std::forward<Types>(ts)...) {}
+
+  /// Returns the number of columns in this row.
+  constexpr std::size_t size() const { return sizeof...(Types); }
+
+  /**
+   *  Returns the value at position `I`.
+   *
+   *  Example:
+   *
+   *      auto row = MakeRow(true, "foo");
+   *      assert(row.get<0>() == true);
+   *      assert(row.get<1>() == "foo");
+   */
+  template <std::size_t I>
+  ColumnType<I> get() const {
+    return std::get<I>(values_);
+  }
+
+  /**
+   * Returns a std::tuple of the values at the specified positions (specified
+   * in any order).
+   *
+   *  Example:
+   *
+   *      auto row = MakeRow(true, "foo", 42);
+   *      std::tuple<std::int64_t, bool> tup = row.get<2, 0>();
+   *      assert(std::get<0>(tup) == 42);
+   *      assert(std::get<1>(tup) == true);
+   */
+  template <std::size_t... I,
+            typename std::enable_if<(sizeof...(I) > 1), int>::type = 0>
+  std::tuple<ColumnType<I>...> get() const {
+    return std::make_tuple(get<I>()...);
+  }
+
+  /**
+   * Returns a std::tuple containing all the values in the row.
+   *
+   *  Example:
+   *
+   *      auto row = MakeRow(true, "foo", 42);
+   *      std::tuple<bool, std::string, std::int64_t> tup
+   *          = row.get();
+   *      assert(std::get<0>(tup) == true);
+   *      assert(std::get<1>(tup) == "foo");
+   *      assert(std::get<2>(tup) == 42);
+   */
+  std::tuple<Types...> get() const { return values_; }
+
+  /// Equality operators.
+  friend bool operator==(Row const& a, Row const& b) {
+    return a.values_ == b.values_;
+  }
+  friend bool operator!=(Row const& a, Row const& b) { return !(a == b); }
```

Comment:
Do we need a `swap` and `std::swap`? And this type has a natural `operator<`, though we can punt on that if you want.

Reply:
I don't know. I find I rarely write swap functions, but maybe I should. What's a good rule of thumb for when we should write a swap vs not?

I filed https://github.com/googleapis/google-cloud-cpp-spanner/issues/112 as a "good first issue" for someone to implement later (maybe me).

Reply:
I guess if we can do a better job than `std::swap` we should implement our own, and I do not see how we could do better in this case.